packages/ingest/src/granule.ts

```diff
@@ -1,71 +1,109 @@
-  );
+  * rename s3 file with timestamp
+  *
+  * @param {string} bucket - bucket of the file
+  * @param {string} key - s3 key of the file
+  * @returns {Promise} promise that resolves when file is renamed
+  */
+export async function renameS3FileWithTimestamp(
+  bucket: string,
+  key: string
+): Promise<void> {
+  const formatString = 'YYYYMMDDTHHmmssSSS';
+  const timestamp = (await S3.headObject(bucket, key)).LastModified;
 
-  const CopySource = encodeurl(`${source.Bucket}/${source.Key}`);
+  if (!timestamp) {
+    throw new Error(`s3://${bucket}/${key} does not have a LastModified property`);
+  }
 
-  const params = {
-    CopySource,
-    Bucket: target.Bucket,
-    Key: target.Key,
-    ...options
-  };
+  let renamedKey = `${key}.v${moment.utc(timestamp).format(formatString)}`;
 
-  return S3.s3CopyObject(params)
-    .catch((error) => {
-      log.error(`Failed to copy s3://${CopySource} to s3://${target.Bucket}/${target.Key}: ${error.message}`);
-      throw error;
-    });
+  // if the renamed file already exists, get a new name
+  // eslint-disable-next-line no-await-in-loop
+  while (await S3.s3ObjectExists({ Bucket: bucket, Key: renamedKey })) {
+    renamedKey = `${key}.v${moment.utc(timestamp).add(1, 'milliseconds').format(formatString)}`;
+  }
+
+  log.debug(`renameS3FileWithTimestamp renaming ${bucket} ${key} to ${renamedKey}`);
+
+  await S3.moveObject({
+    sourceBucket: bucket,
+    sourceKey: key,
+    destinationBucket: bucket,
+    destinationKey: renamedKey,
+    copyTags: true
+  });
 }
 
 /**
-* Move granule file from one s3 bucket & keypath to another
-*
-* @param {Object} source - source
-* @param {string} source.Bucket - source
-* @param {string} source.Key - source
-* @param {Object} target - target
-* @param {string} target.Bucket - target
-* @param {string} target.Key - target
-* @param {Object} options - optional object with properties as defined by AWS API:
-* https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#copyObject-prop
-* @returns {Promise} returns a promise that is resolved when the file is moved
-**/
-async function moveGranuleFile(source, target, options) {
-  deprecate(
-    '@cumulus/ingest/granule.moveGranuleFile',
-    '1.22.1',
-    '@cumulus/aws-client/S3.moveObject'
-  );
+  * get all renamed s3 files for a given bucket and key
+  *
+  * @param {string} bucket - bucket of the file
+  * @param {string} key - s3 key of the file
+  * @returns {Array<Object>} returns renamed files
+  */
+export async function getRenamedS3File(
+  bucket: string,
+  key: string
+): Promise<{ Bucket: string, Key: string, size: number }[]> {
+  const s3list = await S3.listS3ObjectsV2({
+    Bucket: bucket,
+    Prefix: `${key}.v`
+  });
 
-  await copyGranuleFile(source, target, options);
-  return S3.deleteS3Object(source.Bucket, source.Key);
+  if (s3list === undefined) {
+    return [];
+  }
+
+  return s3list
+    .filter((o) => o.Key !== undefined && o.Size !== undefined)
```

Comment:
Would it be better if `listS3ObjectsV2` itself handled these conditions, so it always returns at least an empty array with any objects with undefined bucket/key removed? Seems like it might be be better than repeating this logic everywhere we call the function

Reply:
See [https://github.com/nasa/cumulus/pull/1798/commits/08ceba70690ea61f11cecfef5c6a71d1e1c13c47 ](https://github.com/nasa/cumulus/pull/1798/commits/08ceba70690ea61f11cecfef5c6a71d1e1c13c47)#diff-69e435f10b5478031e5d601d73639322